pkg/autoops/domain/auto_ops_rule.go

```diff
@@ -95,47 +139,129 @@ func (a *AutoOpsRule) SetOpsType(opsType proto.OpsType) {
 	a.AutoOpsRule.TriggeredAt = 0
 }
 
+func (a *AutoOpsRule) SetAutoOpsStatus(status proto.AutoOpsStatus) {
+	a.AutoOpsRule.AutoOpsStatus = status
+	a.AutoOpsRule.UpdatedAt = time.Now().Unix()
+}
+
 func (a *AutoOpsRule) AddOpsEventRateClause(oerc *proto.OpsEventRateClause) (*proto.Clause, error) {
 	ac, err := ptypes.MarshalAny(oerc)
 	if err != nil {
 		return nil, err
 	}
-	return a.addClause(ac)
+	clause, err := a.addClause(ac, oerc.ActionType)
+	if err != nil {
+		return nil, err
+	}
+	a.AutoOpsRule.UpdatedAt = time.Now().Unix()
+	a.AutoOpsRule.TriggeredAt = 0
+	return clause, nil
 }
 
 func (a *AutoOpsRule) AddDatetimeClause(dc *proto.DatetimeClause) (*proto.Clause, error) {
 	ac, err := ptypes.MarshalAny(dc)
 	if err != nil {
 		return nil, err
 	}
-	return a.addClause(ac)
+	var insertIndex = int64(-1)
```

Comment:
You can use `sort.Slice` for this case. It will be cleaner and safer.

```go
	sort.Slice(clauses, func(i, j int) bool {
		return clauses[i].Time < clauses[j].Time
	})
```
Append the new clause as usual, then you unmarshal them to sort it.